lib/schedule/hastus/activity.ex

```diff
@@ -75,4 +82,352 @@ defmodule Schedule.Hastus.Activity do
+              "wad" -> :wad
+              "rad" -> :rad
+            end,
+          start_time: activity.start_time,
+          end_time: activity.end_time,
+          start_place: activity.start_place,
+          end_place: activity.end_place
+        }
+      ],
+      end_time: activity.end_time,
+      end_place: activity.end_place
+    }
+  end
+
+  @spec add_deadheads_to_pieces([__MODULE__.t() | Piece.t()]) :: [__MODULE__.t() | Piece.t()]
+  defp add_deadheads_to_pieces([]) do
+    []
+  end
+
+  defp add_deadheads_to_pieces([
+         %__MODULE__{activity_type: "Deadhead from"} = deadhead,
+         %Piece{} = piece
+         | rest
+       ]) do
+    dummy_trip = Schedule.Trip.from_leading_deadhead(deadhead, piece.block_id)
+
+    new_piece = %{
+      piece
+      | start_time: deadhead.start_time,
+        start_place: deadhead.start_place,
+        trips: [dummy_trip | piece.trips]
+    }
+
+    add_deadheads_to_pieces([new_piece | rest])
+  end
+
+  defp add_deadheads_to_pieces([
+         %Piece{} = piece,
+         %__MODULE__{activity_type: "Deadhead to"} = deadhead
+         | rest
+       ]) do
+    dummy_trip = Schedule.Trip.from_following_deadhead(deadhead, piece.block_id)
+
+    new_piece = %{
+      piece
+      | trips: piece.trips ++ [dummy_trip],
+        end_time: deadhead.end_time,
+        end_place: deadhead.end_place
+    }
+
+    add_deadheads_to_pieces([new_piece | rest])
+  end
+
+  defp add_deadheads_to_pieces([first | rest]) do
+    [first | add_deadheads_to_pieces(rest)]
+  end
+
+  @spec add_sign_ons_to_pieces([__MODULE__.t() | Piece.t()]) :: [__MODULE__.t() | Piece.t()]
+  defp add_sign_ons_to_pieces([]) do
+    []
+  end
+
+  defp add_sign_ons_to_pieces([
+         %__MODULE__{activity_type: "Sign-on"} = sign_on,
+         %Piece{} = piece
+         | rest
+       ]) do
+    new_piece = %{
+      piece
+      | start_time: sign_on.start_time,
+        start_place: sign_on.start_place
+    }
+
+    add_sign_ons_to_pieces([new_piece | rest])
+  end
+
+  defp add_sign_ons_to_pieces([first | rest]) do
+    [first | add_sign_ons_to_pieces(rest)]
+  end
+
+  @spec activities_to_breaks([__MODULE__.t() | Piece.t()]) :: [Break.t() | Piece.t()]
+  defp activities_to_breaks(activities_and_pieces) do
+    Enum.map(activities_and_pieces, fn
+      %Piece{} = piece -> piece
+      %__MODULE__{} = activity -> Break.from_activity(activity)
+    end)
+  end
+
+  @spec trip_in_operator(__MODULE__.t(), Hastus.Trip.t()) :: boolean()
+  defp trip_in_operator(%__MODULE__{activity_type: "Operator"} = activity, trip) do
+    String.contains?(trip.block_id, activity.partial_block_id) and
+      trip.start_time >= activity.start_time and
+      trip.start_time <= activity.end_time
+  end
+
+  @spec operator_is_as_directed?(__MODULE__.t()) :: boolean()
+  defp operator_is_as_directed?(%__MODULE__{activity_type: "Operator"} = activity) do
+    String.contains?(activity.partial_block_id, "ad")
```

Comment:
Would it make sense to just explicitly test for `rad` or `wad` rather than doing this "match anything that contains `ad`" logic?